File: internal/frozen/frozen.go
package frozen

import (
	"archive/tar"
	"compress/gzip"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/BrianNguyen29/x-harness/internal/schema"
)

var includePaths = []string{
	"README.md",
	"AGENTS.md",
	"X_HARNESS.md",
	"CHANGELOG.md",
	"LICENSE",
	"docs",
	"schemas",
	"policies",
	"templates",
	"adapters",
	"components/registry.yaml",
	"examples/golden",
	"examples/adversarial",
	"tools/experimental/evolve",
}

// FrozenManifestFile represents a single file in the manifest.
type FrozenManifestFile struct {
	Path   string `json:"path"`
	Sha256 string `json:"sha256"`
	Size   int    `json:"size"`
}

// FrozenManifest represents the bundle manifest.
type FrozenManifest struct {
	SchemaVersion        string              `json:"schema_version"`
	BundleID             string              `json:"bundle_id"`
	XHarnessVersion      string              `json:"x_harness_version"`
	CreatedAt            string              `json:"created_at"`
	SourceCommit         string              `json:"source_commit"`
	MaturityLevel        string              `json:"maturity_level"`
	Benchmark            map[string]any      `json:"benchmark"`
	IncludedComponents   []string            `json:"included_components"`
	Files                []FrozenManifestFile `json:"files"`
	Signing              map[string]any      `json:"signing"`
}

// FrozenVerifyResult is the result of verify.
type FrozenVerifyResult struct {
	OK          bool           `json:"ok"`
	BundlePath  string         `json:"bundle_path"`
	Manifest    *FrozenManifest `json:"manifest"`
	FileCount   int            `json:"file_count"`
	Errors      []string       `json:"errors"`
}

// FrozenImportResult is the result of import.
type FrozenImportResult struct {
	OK        bool     `json:"ok"`
	DryRun    bool     `json:"dry_run"`
	Target    string   `json:"target"`
	Planned   []string `json:"planned"`
	Written   []string `json:"written"`
	Skipped   []string `json:"skipped"`
	Conflicts []string `json:"conflicts"`
	Errors    []string `json:"errors"`
}

// FrozenExportResult is the result of export.
type FrozenExportResult struct {
	OK        bool           `json:"ok"`
	Out       string         `json:"out"`
	Manifest  *FrozenManifest `json:"manifest"`
	FileCount int            `json:"file_count"`
}

// BundleEntry represents a single entry in the bundle.
type BundleEntry struct {
	Path string
	Data []byte
}

func sha256Buffer(data []byte) string {
	h := sha256.New()
	h.Write(data)
	return fmt.Sprintf("%x", h.Sum(nil))
}

func assertSafeArchivePath(relativePath string) error {
	normalized := filepath.ToSlash(filepath.Clean(relativePath))
	if strings.HasPrefix(normalized, "../") || normalized == ".." || filepath.IsAbs(normalized) || normalized != relativePath {
		return fmt.Errorf("unsafe archive path: %s", relativePath)
	}
	return nil
}

func collectFiles(root, relativePath string) ([]string, error) {
	full := filepath.Join(root, relativePath)
	info, err := os.Stat(full)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	if !info.IsDir() {
		return []string{filepath.ToSlash(relativePath)}, nil
	}
	entries, err := os.ReadDir(full)
	if err != nil {
		return nil, err
	}
	var files []string
	for _, entry := range entries {
		child := filepath.Join(relativePath, entry.Name())
		childFiles, err := collectFiles(root, child)
		if err != nil {
			return nil, err
		}
		files = append(files, childFiles...)
	}
	sort.Strings(files)
	return files, nil
}

func gitCommit(root string) string {
	cmd := exec.Command("git", "rev-parse", "HEAD")
	cmd.Dir = root
	out, err := cmd.Output()
	if err != nil {
		return "unknown"
	}
	return strings.TrimSpace(string(out))
}

func packageVersion(root string) string {
	pkgPath := filepath.Join(root, "packages", "cli", "package.json")
	data, err := os.ReadFile(pkgPath)
	if err != nil {
		return "unknown"
	}
	var pkg struct {
		Version string `json:"version"`
	}
	if err := json.Unmarshal(data, &pkg); err != nil {
		return "unknown"
	}
	if pkg.Version != "" {
		return pkg.Version
	}
	return "unknown"
}

func componentIds(root string) ([]string, error) {
	registryPath := filepath.Join(root, "components", "registry.yaml")
	data, err := os.ReadFile(registryPath)
	if err != nil {
		return nil, nil
	}
	re := regexp.MustCompile(`^\s*-\s+id:\s+(.+)$`)
	lines := strings.Split(string(data), "\n")
	var ids []string
	for _, line := range lines {
		m := re.FindStringSubmatch(line)
		if len(m) > 1 {
			ids = append(ids, strings.TrimSpace(m[1]))
		}
	}
	return ids, nil
}

func buildManifest(root string, files []FrozenManifestFile) (*FrozenManifest, error) {
	now := time.Now().UTC()
	bundleID := fmt.Sprintf("xh_frozen_%s", now.Format("20060102150405"))
	components, err := componentIds(root)
	if err != nil {
		return nil, err
	}
	return &FrozenManifest{
		SchemaVersion:   "1",
		BundleID:        bundleID,
		XHarnessVersion: packageVersion(root),
		CreatedAt:       now.Format(time.RFC3339Nano),
		SourceCommit:    gitCommit(root),
		MaturityLevel:   "H2",
		Benchmark: map[string]any{
			"false_accept_count":           0,
			"adversarial_false_accept_count": 0,
			"episode_packaging_success_rate": nil,
		},
		IncludedComponents: components,
		Files:              files,
		Signing: map[string]any{
			"mode": "unsigned",
		},
	}, nil
}

func validateManifest(manifest *FrozenManifest, root string) ([]string, error) {
	schemaPath := filepath.Join(root, "schemas", "frozen-manifest.schema.json")
	v, err := schema.Compile(schemaPath)
	if err != nil {
		return nil, err
	}
	// jsonschema/v6 requires JSON-compatible input, not a Go struct pointer.
	data, err := json.Marshal(manifest)
	if err != nil {
		return []string{err.Error()}, nil
	}
	var doc any
	if err := json.Unmarshal(data, &doc); err != nil {
		return []string{err.Error()}, nil
	}
	if err := v.Validate(doc); err != nil {
		return []string{err.Error()}, nil
	}
	return nil, nil
}

func createTarGz(entries []BundleEntry) ([]byte, error) {
	var buf strings.Builder
	gw := gzip.NewWriter(&buf)
	tw := tar.NewWriter(gw)
	for _, entry := range entries {
		header := &tar.Header{
			Name: entry.Path,
			Mode: 0644,
			Size: int64(len(entry.Data)),
		}
		if err := tw.WriteHeader(header); err != nil {
			return nil, err
		}
		if _, err := tw.Write(entry.Data); err != nil {
			return nil, err
		}
	}
	if err := tw.Close(); err != nil {
		return nil, err
	}
	if err := gw.Close(); err != nil {
		return nil, err
	}
	return []byte(buf.String()), nil
}

func readTarGz(data []byte) ([]BundleEntry, error) {
	gr, err := gzip.NewReader(strings.NewReader(string(data)))
	if err != nil {
		return nil, err
	}
	defer gr.Close()
	tr := tar.NewReader(gr)
	var entries []BundleEntry
	for {
		header, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		if header.Typeflag != tar.TypeReg && header.Typeflag != 0 {
			continue
		}
		data := make([]byte, header.Size)
		_, err = io.ReadFull(tr, data)
		if err != nil {
			return nil, err
		}
		entries = append(entries, BundleEntry{Path: header.Name, Data: data})
	}
	return entries, nil
}

func checksumsFromEntry(entry *BundleEntry) map[string]string {
	checksums := make(map[string]string)
	if entry == nil {
		return checksums
	}
	lines := strings.Split(string(entry.Data), "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.SplitN(line, "  ", 2)
		if len(parts) == 2 {
			checksums[parts[1]] = parts[0]
		}
	}
	return checksums
}

// ExportFrozenBundle exports a frozen bundle.
func ExportFrozenBundle(root, out string) (*FrozenExportResult, error) {
	root, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	out, err = filepath.Abs(out)
	if err != nil {
		return nil, err
	}

	var relativeFiles []string
	for _, item := range includePaths {
		files, err := collectFiles(root, item)
		if err != nil {
			return nil, err
		}
		relativeFiles = append(relativeFiles, files...)
	}
	sort.Strings(relativeFiles)

	var payloadEntries []BundleEntry
	var manifestFiles []FrozenManifestFile
	for _, relativePath := range relativeFiles {
		if err := assertSafeArchivePath(relativePath); err != nil {
			return nil, err
		}
		data, err := os.ReadFile(filepath.Join(root, relativePath))
		if err != nil {
			return nil, err
		}
		manifestFiles = append(manifestFiles, FrozenManifestFile{
			Path:   relativePath,
			Sha256: sha256Buffer(data),
			Size:   len(data),
		})
		payloadEntries = append(payloadEntries, BundleEntry{Path: relativePath, Data: data})
	}

	manifest, err := buildManifest(root, manifestFiles)
	if err != nil {
		return nil, err
	}
	manifestErrors, err := validateManifest(manifest, root)
	if err != nil {
		return nil, err
	}
	if len(manifestErrors) > 0 {
		return nil, fmt.Errorf("frozen manifest validation failed: %s", strings.Join(manifestErrors, "; "))
	}

	checksums := make([]string, 0, len(manifest.Files))
	for _, file := range manifest.Files {
		checksums = append(checksums, fmt.Sprintf("%s  %s", file.Sha256, file.Path))
	}
	version := map[string]any{
		"x_harness_version": manifest.XHarnessVersion,
		"source_commit":     manifest.SourceCommit,
		"created_at":        manifest.CreatedAt,
	}

	manifestJSON, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return nil, err
	}
	versionJSON, err := json.MarshalIndent(version, "", "  ")
	if err != nil {
		return nil, err
	}

	entries := []BundleEntry{
		{Path: "manifest.json", Data: manifestJSON},
		{Path: "checksums.sha256", Data: []byte(strings.Join(checksums, "\n") + "\n")},
		{Path: "version.json", Data: versionJSON},
	}
	entries = append(entries, payloadEntries...)

	bundleData, err := createTarGz(entries)
	if err != nil {
		return nil, err
	}

	if err := os.MkdirAll(filepath.Dir(out), 0755); err != nil {
		return nil, err
	}
	if err := os.WriteFile(out, bundleData, 0644); err != nil {
		return nil, err
	}

	return &FrozenExportResult{
		OK:        true,
		Out:       out,
		Manifest:  manifest,
		FileCount: len(manifest.Files),
	}, nil
}

func readFrozenArchive(bundlePath string) ([]BundleEntry, *FrozenManifest, []BundleEntry, []string, error) {
	data, err := os.ReadFile(bundlePath)
	if err != nil {
		return nil, nil, nil, nil, err
	}
	entries, err := readTarGz(data)
	if err != nil {
		return nil, nil, nil, nil, err
	}
	var errors []string
	var manifestEntry *BundleEntry
	var checksumEntry *BundleEntry
	var payload []BundleEntry
	for i := range entries {
		switch entries[i].Path {
		case "manifest.json":
			manifestEntry = &entries[i]
		case "checksums.sha256":
			checksumEntry = &entries[i]
		case "version.json":
			// metadata, skip
		default:
			payload = append(payload, entries[i])
		}
	}
	if manifestEntry == nil {
		return nil, nil, nil, nil, fmt.Errorf("frozen bundle missing manifest.json")
	}
	var manifest FrozenManifest
	if err := json.Unmarshal(manifestEntry.Data, &manifest); err != nil {
		return nil, nil, nil, nil, fmt.Errorf("invalid manifest.json: %w", err)
	}
	schemaPath := filepath.Join(".", "schemas", "frozen-manifest.schema.json")
	if _, statErr := os.Stat(schemaPath); statErr == nil {
		manifestErrors, err := validateManifest(&manifest, ".")
		if err != nil {
			return nil, nil, nil, nil, err
		}
		errors = append(errors, manifestErrors...)
	}
	checksums := checksumsFromEntry(checksumEntry)

	manifestPaths := make(map[string]struct{})
	checksumPaths := make(map[string]struct{})
	payloadByPath := make(map[string]BundleEntry)

	for _, file := range manifest.Files {
		if err := assertSafeArchivePath(file.Path); err != nil {
			errors = append(errors, err.Error())
			continue
		}
		if _, exists := manifestPaths[file.Path]; exists {
			errors = append(errors, fmt.Sprintf("duplicate manifest file path: %s", file.Path))
			continue
		}
		manifestPaths[file.Path] = struct{}{}
	}

	for _, entry := range payload {
		if err := assertSafeArchivePath(entry.Path); err != nil {
			errors = append(errors, err.Error())
			continue
		}
		if _, exists := payloadByPath[entry.Path]; exists {
			errors = append(errors, fmt.Sprintf("duplicate payload file path: %s", entry.Path))
			continue
		}
		payloadByPath[entry.Path] = entry
		if _, ok := manifestPaths[entry.Path]; !ok {
			errors = append(errors, fmt.Sprintf("payload file not declared in manifest: %s", entry.Path))
		}
	}

	for path := range checksums {
		checksumPaths[path] = struct{}{}
		if _, ok := manifestPaths[path]; !ok {
			errors = append(errors, fmt.Sprintf("checksums.sha256 path not declared in manifest: %s", path))
		}
	}

	var verifiedPayload []BundleEntry
	for _, file := range manifest.Files {
		entry, ok := payloadByPath[file.Path]
		if !ok {
			errors = append(errors, fmt.Sprintf("manifest file missing from bundle: %s", file.Path))
			continue
		}
		actual := sha256Buffer(entry.Data)
		if actual != file.Sha256 {
			errors = append(errors, fmt.Sprintf("checksum mismatch for %s", file.Path))
		}
		if checksums[file.Path] != file.Sha256 {
			errors = append(errors, fmt.Sprintf("checksums.sha256 mismatch for %s", file.Path))
		}
		verifiedPayload = append(verifiedPayload, entry)
	}

	return entries, &manifest, verifiedPayload, errors, nil
}

// VerifyFrozenBundle verifies a frozen bundle.
func VerifyFrozenBundle(bundlePath string) (*FrozenVerifyResult, error) {
	resolved, err := filepath.Abs(bundlePath)
	if err != nil {
		return &FrozenVerifyResult{
			OK:         false,
			BundlePath: bundlePath,
			Manifest:   nil,
			FileCount:  0,
			Errors:     []string{err.Error()},
		}, nil
	}
	_, manifest, payload, errors, err := readFrozenArchive(resolved)
	if err != nil {
		return &FrozenVerifyResult{
			OK:         false,
			BundlePath: resolved,
			Manifest:   nil,
			FileCount:  0,
			Errors:     []string{err.Error()},
		}, nil
	}
	return &FrozenVerifyResult{
		OK:         len(errors) == 0,
		BundlePath: resolved,
		Manifest:   manifest,
		FileCount:  len(payload),
		Errors:     errors,
	}, nil
}

// ImportFrozenBundle imports a frozen bundle.
func ImportFrozenBundle(bundlePath, target string, dryRun, merge, force bool) (*FrozenImportResult, error) {
	resolvedBundle, err := filepath.Abs(bundlePath)
	if err != nil {
		return nil, err
	}
	target, err = filepath.Abs(target)
	if err != nil {
		return nil, err
	}

	_, _, payload, errors, err := readFrozenArchive(resolvedBundle)
	if err != nil {
		return &FrozenImportResult{
			OK:      false,
			DryRun:  dryRun,
			Target:  target,
			Errors:  []string{err.Error()},
		}, nil
	}

	planned := make([]string, 0)
	written := make([]string, 0)
	skipped := make([]string, 0)
	conflicts := make([]string, 0)
	resultErrors := make([]string, 0, len(errors))
	resultErrors = append(resultErrors, errors...)

	for _, entry := range payload {
		if err := assertSafeArchivePath(entry.Path); err != nil {
			resultErrors = append(resultErrors, err.Error())
			continue
		}
		dest := filepath.Join(target, filepath.FromSlash(entry.Path))
		destClean := filepath.Clean(dest)
		targetClean := filepath.Clean(target)
		rel, err := filepath.Rel(targetClean, destClean)
		if err != nil || strings.HasPrefix(rel, "..") {
			resultErrors = append(resultErrors, fmt.Sprintf("unsafe import path: %s", entry.Path))
			continue
		}
		planned = append(planned, entry.Path)
		exists := false
		if _, err := os.Stat(dest); err == nil {
			exists = true
		}
		if exists && merge {
			skipped = append(skipped, entry.Path)
			continue
		}
		if exists && !force && !dryRun {
			conflicts = append(conflicts, entry.Path)
			continue
		}
		if !dryRun {
			if err := os.MkdirAll(filepath.Dir(dest), 0755); err != nil {
				resultErrors = append(resultErrors, err.Error())
				continue
			}
			if err := os.WriteFile(dest, entry.Data, 0644); err != nil {
				resultErrors = append(resultErrors, err.Error())
				continue
			}
			written = append(written, entry.Path)
		}
	}

	if len(conflicts) > 0 {
		resultErrors = append(resultErrors, "protected files already exist; use --merge or --force")
	}

	return &FrozenImportResult{
		OK:        len(resultErrors) == 0,
		DryRun:    dryRun,
		Target:    target,
		Planned:   planned,
		Written:   written,
		Skipped:   skipped,
		Conflicts: conflicts,
		Errors:    resultErrors,
	}, nil
}

File: packages/cli/src/core/frozen.ts
import { execFile } from "node:child_process";
import { createHash } from "node:crypto";
import * as path from "node:path";
import { gzipSync, gunzipSync } from "node:zlib";
import fs from "fs-extra";
import { compileSchema, loadSchema } from "./schema.js";

interface TarEntry {
  path: string;
  data: Buffer;
}

export interface FrozenManifestFile {
  path: string;
  sha256: string;
  size: number;
}

export interface FrozenManifest {
  schema_version: string;
  bundle_id: string;
  x_harness_version: string;
  created_at: string;
  source_commit: string;
  maturity_level: string;
  benchmark: {
    false_accept_count: number;
    adversarial_false_accept_count: number;
    episode_packaging_success_rate: number | null;
  };
  included_components: string[];
  files: FrozenManifestFile[];
  signing: {
    mode: "unsigned" | "sigstore";
    signature_ref?: string;
  };
}

export interface FrozenVerifyResult {
  ok: boolean;
  bundle_path: string;
  manifest: FrozenManifest | null;
  file_count: number;
  errors: string[];
}

export interface FrozenImportResult {
  ok: boolean;
  dry_run: boolean;
  target: string;
  planned: string[];
  written: string[];
  skipped: string[];
  conflicts: string[];
  errors: string[];
}

const INCLUDE_PATHS = [
  "README.md",
  "AGENTS.md",
  "X_HARNESS.md",
  "CHANGELOG.md",
  "LICENSE",
  "docs",
  "schemas",
  "policies",
  "templates",
  "adapters",
  "components/registry.yaml",
  "examples/golden",
  "examples/adversarial",
  "tools/experimental/evolve",
];

function sha256Buffer(data: Buffer): string {
  return createHash("sha256").update(data).digest("hex");
}

async function collectFiles(
  root: string,
  relativePath: string
): Promise<string[]> {
  const full = path.join(root, relativePath);
  if (!(await fs.pathExists(full))) return [];
  const stat = await fs.stat(full);
  if (stat.isFile()) return [relativePath.replaceAll("\\", "/")];
  if (!stat.isDirectory()) return [];
  const entries = await fs.readdir(full, { withFileTypes: true });
  const files: string[] = [];
  for (const entry of entries) {
    const child = path.join(relativePath, entry.name).replaceAll("\\", "/");
    files.push(...(await collectFiles(root, child)));
  }
  return files.sort();
}

async function gitCommit(root: string): Promise<string> {
  return new Promise((resolve) => {
    execFile("git", ["rev-parse", "HEAD"], { cwd: root }, (error, stdout) => {
      if (error) resolve("unknown");
      else resolve(stdout.trim() || "unknown");
    });
  });
}

async function packageVersion(root: string): Promise<string> {
  const packagePath = path.join(root, "packages", "cli", "package.json");
  if (!(await fs.pathExists(packagePath))) return "unknown";
  const pkg = (await fs.readJson(packagePath)) as { version?: string };
  return pkg.version ?? "unknown";
}

async function componentIds(root: string): Promise<string[]> {
  const registryPath = path.join(root, "components", "registry.yaml");
  if (!(await fs.pathExists(registryPath))) return [];
  const content = await fs.readFile(registryPath, "utf-8");
  return [...content.matchAll(/^\s*-\s+id:\s+(.+)$/gm)]
    .map((match) => match[1]?.trim())
    .filter((value): value is string => Boolean(value));
}

function splitTarName(name: string): { name: string; prefix: string } {
  if (Buffer.byteLength(name) <= 100) return { name, prefix: "" };
  const parts = name.split("/");
  for (let i = 1; i < parts.length; i += 1) {
    const prefix = parts.slice(0, i).join("/");
    const shortName = parts.slice(i).join("/");
    if (
      Buffer.byteLength(prefix) <= 155 &&
      Buffer.byteLength(shortName) <= 100
    ) {
      return { name: shortName, prefix };
    }
  }
  throw new Error(`tar path too long: ${name}`);
}

function writeString(
  header: Buffer,
  offset: number,
  length: number,
  value: string
): void {
  header.write(value.slice(0, length), offset, length, "utf-8");
}

function writeOctal(
  header: Buffer,
  offset: number,
  length: number,
  value: number
): void {
  const encoded = value
    .toString(8)
    .padStart(length - 1, "0")
    .slice(0, length - 1);
  writeString(header, offset, length - 1, encoded);
  header[offset + length - 1] = 0;
}

function createTarHeader(entry: TarEntry): Buffer {
  const header = Buffer.alloc(512, 0);
  const names = splitTarName(entry.path);
  writeString(header, 0, 100, names.name);
  writeOctal(header, 100, 8, 0o644);
  writeOctal(header, 108, 8, 0);
  writeOctal(header, 116, 8, 0);
  writeOctal(header, 124, 12, entry.data.length);
  writeOctal(header, 136, 12, Math.floor(Date.now() / 1000));
  header.fill(0x20, 148, 156);
  writeString(header, 156, 1, "0");
  writeString(header, 257, 6, "ustar");
  writeString(header, 263, 2, "00");
  writeString(header, 345, 155, names.prefix);
  const checksum = header.reduce((sum, value) => sum + value, 0);
  const encoded = checksum.toString(8).padStart(6, "0");
  writeString(header, 148, 6, encoded);
  header[154] = 0;
  header[155] = 0x20;
  return header;
}

function pad512(data: Buffer): Buffer {
  const remainder = data.length % 512;
  if (remainder === 0) return data;
  return Buffer.concat([data, Buffer.alloc(512 - remainder, 0)]);
}

function createTarGz(entries: TarEntry[]): Buffer {
  const chunks: Buffer[] = [];
  for (const entry of entries) {
    chunks.push(createTarHeader(entry), pad512(entry.data));
  }
  chunks.push(Buffer.alloc(1024, 0));
  return gzipSync(Buffer.concat(chunks));
}

function parseOctal(value: Buffer): number {
  const text = value.toString("utf-8").replace(/\0/g, "").trim();
  return text ? Number.parseInt(text, 8) : 0;
}

function readTarGz(data: Buffer): TarEntry[] {
  const tar = gunzipSync(data);
  const entries: TarEntry[] = [];
  let offset = 0;
  while (offset + 512 <= tar.length) {
    const header = tar.subarray(offset, offset + 512);
    if (header.every((value) => value === 0)) break;
    const name = header.subarray(0, 100).toString("utf-8").replace(/\0.*$/, "");
    const prefix = header
      .subarray(345, 500)
      .toString("utf-8")
      .replace(/\0.*$/, "");
    const size = parseOctal(header.subarray(124, 136));
    const typeflag = header.subarray(156, 157).toString("utf-8");
    const fullName = prefix ? `${prefix}/${name}` : name;
    offset += 512;
    const body = tar.subarray(offset, offset + size);
    if (typeflag === "0" || typeflag === "\0") {
      entries.push({ path: fullName, data: Buffer.from(body) });
    }
    offset += Math.ceil(size / 512) * 512;
  }
  return entries;
}

function assertSafeArchivePath(relativePath: string): void {
  const normalized = path.posix.normalize(relativePath);
  if (
    normalized.startsWith("../") ||
    normalized === ".." ||
    path.posix.isAbsolute(normalized) ||
    normalized !== relativePath
  ) {
    throw new Error(`unsafe archive path: ${relativePath}`);
  }
}

async function buildManifest(
  root: string,
  files: FrozenManifestFile[]
): Promise<FrozenManifest> {
  return {
    schema_version: "1",
    bundle_id: `xh_frozen_${new Date()
      .toISOString()
      .replace(/[-:.TZ]/g, "")
      .slice(0, 14)}`,
    x_harness_version: await packageVersion(root),
    created_at: new Date().toISOString(),
    source_commit: await gitCommit(root),
    maturity_level: "H2",
    benchmark: {
      false_accept_count: 0,
      adversarial_false_accept_count: 0,
      episode_packaging_success_rate: null,
    },
    included_components: await componentIds(root),
    files,
    signing: {
      mode: "unsigned",
    },
  };
}

async function validateManifest(manifest: FrozenManifest): Promise<string[]> {
  const schema = await loadSchema("frozen-manifest");
  const validate = compileSchema(schema);
  if (validate(manifest)) return [];
  return (validate.errors ?? []).map(
    (err) => `${err.instancePath || "/"} ${err.message ?? "invalid"}`
  );
}

export async function exportFrozenBundle(input: {
  root: string;
  out: string;
}): Promise<{
  ok: boolean;
  out: string;
  manifest: FrozenManifest;
  file_count: number;
}> {
  const root = path.resolve(input.root);
  const out = path.resolve(input.out);
  const relativeFiles = (
    await Promise.all(INCLUDE_PATHS.map((item) => collectFiles(root, item)))
  )
    .flat()
    .sort();
  const payloadEntries: TarEntry[] = [];
  const manifestFiles: FrozenManifestFile[] = [];
  for (const relativePath of relativeFiles) {
    assertSafeArchivePath(relativePath);
    const data = await fs.readFile(path.join(root, relativePath));
    manifestFiles.push({
      path: relativePath,
      sha256: sha256Buffer(data),
      size: data.length,
    });
    payloadEntries.push({ path: relativePath, data });
  }
  const manifest = await buildManifest(root, manifestFiles);
  const manifestErrors = await validateManifest(manifest);
  if (manifestErrors.length > 0) {
    throw new Error(
      `frozen manifest validation failed: ${manifestErrors.join("; ")}`
    );
  }
  const checksums = manifest.files
    .map((file) => `${file.sha256}  ${file.path}`)
    .join("\n");
  const version = {
    x_harness_version: manifest.x_harness_version,
    source_commit: manifest.source_commit,
    created_at: manifest.created_at,
  };
  const entries: TarEntry[] = [
    {
      path: "manifest.json",
      data: Buffer.from(JSON.stringify(manifest, null, 2), "utf-8"),
    },
    {
      path: "checksums.sha256",
      data: Buffer.from(`${checksums}\n`, "utf-8"),
    },
    {
      path: "version.json",
      data: Buffer.from(JSON.stringify(version, null, 2), "utf-8"),
    },
    ...payloadEntries,
  ];
  await fs.ensureDir(path.dirname(out));
  await fs.writeFile(out, createTarGz(entries));
  return { ok: true, out, manifest, file_count: manifest.files.length };
}

function checksumsFromEntry(entry: TarEntry | undefined): Map<string, string> {
  const checksums = new Map<string, string>();
  if (!entry) return checksums;
  const lines = entry.data.toString("utf-8").split(/\r?\n/).filter(Boolean);
  for (const line of lines) {
    const match = /^([a-f0-9]{64})\s+(.+)$/.exec(line);
    if (match) checksums.set(match[2], match[1]);
  }
  return checksums;
}

async function readFrozenArchive(bundlePath: string): Promise<{
  entries: TarEntry[];
  manifest: FrozenManifest;
  payload: TarEntry[];
  errors: string[];
}> {
  const entries = readTarGz(await fs.readFile(bundlePath));
  const errors: string[] = [];
  const manifestEntry = entries.find((entry) => entry.path === "manifest.json");
  if (!manifestEntry) {
    throw new Error("frozen bundle missing manifest.json");
  }
  const manifest = JSON.parse(
    manifestEntry.data.toString("utf-8")
  ) as FrozenManifest;
  errors.push(...(await validateManifest(manifest)));
  const checksums = checksumsFromEntry(
    entries.find((entry) => entry.path === "checksums.sha256")
  );
  const payload = entries.filter(
    (entry) =>
      !["manifest.json", "checksums.sha256", "version.json"].includes(
        entry.path
      )
  );
  const manifestPaths = new Set<string>();
  const checksumPaths = new Set(checksums.keys());
  const payloadByPath = new Map<string, TarEntry>();

  for (const file of manifest.files) {
    try {
      assertSafeArchivePath(file.path);
    } catch (err) {
      errors.push(err instanceof Error ? err.message : String(err));
      continue;
    }
    if (manifestPaths.has(file.path)) {
      errors.push(`duplicate manifest file path: ${file.path}`);
      continue;
    }
    manifestPaths.add(file.path);
  }

  for (const entry of payload) {
    try {
      assertSafeArchivePath(entry.path);
    } catch (err) {
      errors.push(err instanceof Error ? err.message : String(err));
      continue;
    }
    if (payloadByPath.has(entry.path)) {
      errors.push(`duplicate payload file path: ${entry.path}`);
      continue;
    }
    payloadByPath.set(entry.path, entry);
    if (!manifestPaths.has(entry.path)) {
      errors.push(`payload file not declared in manifest: ${entry.path}`);
    }
  }

  for (const checksumPath of checksumPaths) {
    if (!manifestPaths.has(checksumPath)) {
      errors.push(
        `checksums.sha256 path not declared in manifest: ${checksumPath}`
      );
    }
  }

  for (const file of manifest.files) {
    const entry = payloadByPath.get(file.path);
    if (!entry) {
      errors.push(`manifest file missing from bundle: ${file.path}`);
      continue;
    }
    const actual = sha256Buffer(entry.data);
    if (actual !== file.sha256) {
      errors.push(`checksum mismatch for ${file.path}`);
    }
    if (checksums.get(file.path) !== file.sha256) {
      errors.push(`checksums.sha256 mismatch for ${file.path}`);
    }
  }
  const verifiedPayload = manifest.files
    .map((file) => payloadByPath.get(file.path))
    .filter((entry): entry is TarEntry => Boolean(entry));
  return { entries, manifest, payload: verifiedPayload, errors };
}

export async function verifyFrozenBundle(
  bundlePath: string
): Promise<FrozenVerifyResult> {
  try {
    const resolved = path.resolve(bundlePath);
    const archive = await readFrozenArchive(resolved);
    return {
      ok: archive.errors.length === 0,
      bundle_path: resolved,
      manifest: archive.manifest,
      file_count: archive.payload.length,
      errors: archive.errors,
    };
  } catch (err) {
    return {
      ok: false,
      bundle_path: path.resolve(bundlePath),
      manifest: null,
      file_count: 0,
      errors: [err instanceof Error ? err.message : String(err)],
    };
  }
}

export async function importFrozenBundle(input: {
  bundlePath: string;
  target: string;
  dryRun: boolean;
  merge?: boolean;
  force?: boolean;
}): Promise<FrozenImportResult> {
  const resolvedBundle = path.resolve(input.bundlePath);
  const archive = await readFrozenArchive(resolvedBundle).catch((err) => ({
    entries: [],
    manifest: null as unknown as FrozenManifest,
    payload: [] as TarEntry[],
    errors: [err instanceof Error ? err.message : String(err)],
  }));
  const errors = [...archive.errors];
  const target = path.resolve(input.target);
  const planned: string[] = [];
  const written: string[] = [];
  const skipped: string[] = [];
  const conflicts: string[] = [];
  if (errors.length > 0) {
    return {
      ok: false,
      dry_run: input.dryRun,
      target,
      planned,
      written,
      skipped,
      conflicts,
      errors,
    };
  }

  for (const entry of archive.payload) {
    assertSafeArchivePath(entry.path);
    const dest = path.resolve(target, entry.path);
    if (dest !== target && !dest.startsWith(`${target}${path.sep}`)) {
      errors.push(`unsafe import path: ${entry.path}`);
      continue;
    }
    planned.push(entry.path);
    const exists = await fs.pathExists(dest);
    if (exists && input.merge) {
      skipped.push(entry.path);
      continue;
    }
    if (exists && !input.force && !input.dryRun) {
      conflicts.push(entry.path);
      continue;
    }
    if (!input.dryRun) {
      await fs.ensureDir(path.dirname(dest));
      await fs.writeFile(dest, entry.data);
      written.push(entry.path);
    }
  }

  if (conflicts.length > 0) {
    errors.push("protected files already exist; use --merge or --force");
  }

  return {
    ok: errors.length === 0,
    dry_run: input.dryRun,
    target,
    planned,
    written,
    skipped,
    conflicts,
    errors,
  };
}
